File: Snippets/Haskell/Fig.hs
module Fig where

data FIGDoc = FIGDoc { version          :: Float
                     , orientation      :: FIGOrientation
                     , justification    :: FIGJustification
                     , units            :: FIGUnits
                     , paperSize        :: FIGPaperSize
                     , magnification    :: Float
                     , multiPage        :: FIGMultiPage
                     , transparentColor :: Int
                     , comment          :: String
                     , resolution       :: Int
                     , coordSystem      :: FIGCoordSystem
                     } deriving (Show)

data FIGOrientation   = Landscape | Portrait    deriving (Show)
data FIGJustification = Center    | FlushLeft   deriving (Show)
data FIGUnits         = Metric    | Inches      deriving (Show)
data FIGPaperSize     = Letter    | Legal     |
                        Ledger    | Tabloid   |
                        A         | B         |
                        C         | D         |
                        E         | A4        |
                        A3        | A2        |
                        A1        | A0        |
                        B5                      deriving (Show)
data FIGMultiPage     = Single    | Multiple    deriving (Show)
data FIGCoordSystem   = LowerLeft | UpperLeft   deriving (Show)

-- data FIGObject = Color | 

main :: IO ()
main = interact transFIG

readToken :: String -> String
readToken [] = 

transFIG :: String -> String
transFIG s = show $ parseFIG s

parseFIG :: String -> FIGDoc
parseFIG s = parseVersion s

-- parseFIG x = FIGDoc { version = 3.2
--                     , orientation = Landscape
--                     , justification = Center
--                     , units = Metric
--                     , paperSize = A4
--                     , magnification = 4.5
--                     , multiPage = Single
--                     , transparentColor = 0
--                     , comment = "Foobar"
--                     , resolution = 1200
--                     , coordSystem = LowerLeft
--                     }

File: Projects/VerilogOnline/archive/caso-old-2/fig.py
def parse_document(lines):
    return map(_parse_line, lines, range(len(lines)))

def _parse_line(line, num):
    return _parse_line_helper(line.strip(), num)

def _parse_line_helper(line, num):
    if line.startswith("#"):
        if num == 0:
            return ("version", line.split(" ")[1])
        else:
            return None
    elif num == 0:
        return ("error", line, num,
                "expecting line 0 to start with #FIG <version>")
    elif num == 1:
        if line in ["Landscape", "Portrait"]:
            return ("orientation", line)
        else:
            return ("error", line, num,
                    "line 1 must have an orientation property")
    elif num == 2:
        if line in ["Center", "Flush Left"]:
            return ("justification", line)
        else:
            return ("error", line, num,
                    "line 2 must have a justification property")
    elif num == 3:
        if line in ["Metric", "Inches"]:
            return ("units", line)
        else:
            return ("error", line, num,
                    "line 3 must have a unitrs property")
    elif num == 4:
        if line in ["Letter", "Legal", "Ledger", "Tabloid",
                    "A", "B", "C", "D", "E",
                    "A4", "A3", "A2", "A1", "A0", "B5"]:
            return ("papersize", line)
        else:
            return ("error", line, num,
                    "line 4 must have a papersize property")
    else:
        pass
